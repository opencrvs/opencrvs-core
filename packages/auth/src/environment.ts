/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { cleanEnv, str, port, url, num, bool } from 'envalid'

export const env = cleanEnv(process.env, {
  REDIS_HOST: str({ devDefault: 'localhost' }),
  AUTH_HOST: str({ default: '0.0.0.0' }),
  AUTH_PORT: port({ default: 4040 }),
  USER_MANAGEMENT_URL: url({ devDefault: 'http://localhost:3030/' }),
  METRICS_URL: url({ devDefault: 'http://localhost:1050' }),
  NOTIFICATION_SERVICE_URL: url({ devDefault: 'http://localhost:2020/' }),
  DOMAIN: str({ devDefault: '*' }),
  COUNTRY_CONFIG_URL: url({ devDefault: 'http://localhost:3040/' }), // used for external requests (CORS whitelist)
  COUNTRY_CONFIG_URL_INTERNAL: url({ devDefault: 'http://localhost:3040/' }), // used for internal service-to-service communication
  LOGIN_URL: url({ devDefault: 'http://localhost:3020/' }),
  CLIENT_APP_URL: url({ devDefault: 'http://localhost:3000/' }),
  CERT_PRIVATE_KEY_PATH: str({ devDefault: '../../.secrets/private-key.pem' }),
  CERT_PUBLIC_KEY_PATH: str({ devDefault: '../../.secrets/public-key.pem' }),
  SENTRY_DSN: str({ default: undefined }),
  QA_ENV: bool({ default: false }),

  CONFIG_TOKEN_EXPIRY_SECONDS: num({ default: 604800 }), // 1 week
  CONFIG_SMS_CODE_EXPIRY_SECONDS: num({ default: 600 }), // 10 minutes
  CONFIG_SYSTEM_TOKEN_EXPIRY_SECONDS: num({ default: 600 }) // 10 minutes
})
