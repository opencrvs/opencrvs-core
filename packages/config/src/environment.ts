/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { cleanEnv, str, port, url, bool } from 'envalid'

export const env = cleanEnv(process.env, {
  HOST: str({ devDefault: 'localhost' }),
  DOMAIN: str({ devDefault: '*' }),
  LOGIN_URL: url({ devDefault: 'http://localhost:3020/' }),
  CLIENT_APP_URL: url({ devDefault: 'http://localhost:3000/' }),
  PORT: port({ default: 2021 }),
  QA_ENV: bool({ default: false }),
  CHECK_INVALID_TOKEN: bool({
    devDefault: false,
    desc: `Check if the token has been invalided in the auth service before it has expired`
  }),
  CERT_PUBLIC_KEY_PATH: str({ devDefault: '../../.secrets/public-key.pem' }),
  SENTRY_DSN: str({ default: undefined }),
  AUTH_URL: url({ devDefault: 'http://localhost:4040/' }),
  COUNTRY_CONFIG_URL: url({ devDefault: 'http://localhost:3040/' }),
  MONGO_URL: url({ devDefault: 'mongodb://localhost/application-config' }),
  USER_MANAGEMENT_URL: url({ devDefault: 'http://localhost:3030/' }),
  FHIR_URL: url({ devDefault: 'http://localhost:3447/fhir' }),
  HEARTH_MONGO_URL: str({ default: 'mongodb://localhost/hearth-dev' })
})
