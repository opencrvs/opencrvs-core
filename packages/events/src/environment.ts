/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { cleanEnv, url } from 'envalid'

/**
 * When defining variables aim to be consistent with existing values.
 * Define URLs without trailing slashes, include the protocol.
 */
export const env = cleanEnv(process.env, {
  EVENTS_MONGO_URL: url({ devDefault: 'mongodb://localhost/events' }),
  USER_MGNT_MONGO_URL: url({ devDefault: 'mongodb://localhost/user-mgnt' }),
  ES_URL: url({ devDefault: 'http://localhost:9200' }),
  COUNTRY_CONFIG_URL: url({ devDefault: 'http://localhost:3040' }),
  DOCUMENTS_URL: url({ devDefault: 'http://localhost:9050' }),
  USER_MANAGEMENT_URL: url({ devDefault: 'http://localhost:3030' }),
  AUTH_URL: url({ devDefault: 'http://localhost:4040' })
})
