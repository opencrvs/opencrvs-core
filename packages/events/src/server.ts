/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createServer, IncomingMessage } from 'http'
import { createOpenApiHttpHandler } from 'trpc-to-openapi'
import { createHTTPHandler } from '@trpc/server/adapters/standalone'
import '@opencrvs/commons/monitoring'
import { logger } from '@opencrvs/commons'
import { appRouter } from './router/router'
import { createContext } from './context'

function stringifyRequest(req: IncomingMessage) {
  const url = new URL(req.url || '', `http://${req.headers.host}`)
  return `'${req.method} ${url.pathname}'`
}

const trpcConfig: Parameters<typeof createHTTPHandler>[0] = {
  router: appRouter,
  middleware: (req, _, next) => {
    logger.info(`Request: ${stringifyRequest(req)}`)
    return next()
  },
  onError: ({ error, req }) => {
    logger.warn(
      `Error for request: ${stringifyRequest(req)}. Error: '${error.message}'`
    )
  },
  createContext
}

function isTrpcRequest(req: IncomingMessage): boolean {
  if (!req.url) {
    throw new Error('No URL provided')
  }
  const url = new URL(req.url, `http://${req.headers.host}`)
  const pathName = url.pathname.replace(/^\//, '')
  const trpcProcedurePaths = Object.keys(appRouter._def.procedures)

  return (
    url.search.startsWith('?input') || trpcProcedurePaths.includes(pathName)
  )
}

export function server() {
  const restServer = createOpenApiHttpHandler(trpcConfig)
  const trpcServer = createHTTPHandler(trpcConfig)

  return createServer((req, res) => {
    if (!req.url) {
      res.writeHead(500)
      res.end('No URL provided')
      return
    }

    // If it's a tRPC request, handle it with the tRPC server
    if (isTrpcRequest(req)) {
      trpcServer(req, res)
    } else {
      // If it's a REST request, handle it with the REST server
      void restServer(req, res)
    }
  })
}
