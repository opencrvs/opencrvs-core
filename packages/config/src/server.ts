/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import * as Hapi from '@hapi/hapi'
import { DEFAULT_TIMEOUT } from '@config/config/constants'
import getRoutes from '@config/config/routes'
import getPlugins from '@config/config/plugins'
import * as database from '@config/config/database'
import * as mongoDirect from '@config/config/hearthClient'
import { validateFunc, logger } from '@opencrvs/commons'
import { readFileSync } from 'fs'
import { badRequest } from '@hapi/boom'
import { env } from './environment'

export const publicCert = readFileSync(env.CERT_PUBLIC_KEY_PATH)

export async function createServer() {
  let whitelist: string[] = [env.DOMAIN]
  if (env.DOMAIN[0] !== '*') {
    whitelist = [env.LOGIN_URL, env.CLIENT_APP_URL]
  }
  logger.info(`Whitelist: ${JSON.stringify(whitelist)}`)
  const server = new Hapi.Server({
    host: env.HOST,
    port: env.PORT,
    routes: {
      cors: { origin: whitelist },
      validate: {
        failAction: async (_, _2, err) => {
          if (process.env.NODE_ENV === 'production') {
            // In prod, log a limited error message and throw the default Bad Request error.
            logger.error(`ValidationError: ${err?.message}`)
            throw badRequest(`Invalid request payload input`)
          } else {
            // During development, log and respond with the full error.
            logger.error(err?.message)
            throw err
          }
        }
      },
      payload: { maxBytes: 52428800, timeout: DEFAULT_TIMEOUT }
    }
  })

  await server.register(getPlugins())

  server.auth.strategy('jwt', 'jwt', {
    key: publicCert,
    verifyOptions: {
      algorithms: ['RS256'],
      issuer: 'opencrvs:auth-service',
      audience: 'opencrvs:config-user'
    },
    validate: (payload: any, request: Hapi.Request) =>
      validateFunc(payload, request, env.CHECK_INVALID_TOKEN, env.AUTH_URL)
  })

  server.auth.default('jwt')

  const routes = getRoutes()
  server.route(routes)

  server.ext({
    type: 'onRequest',
    method(request: Hapi.Request & { sentryScope?: any }, h) {
      request.sentryScope?.setExtra('payload', request.payload)
      return h.continue
    }
  })

  async function stop() {
    await server.stop()
    await database.stop()
    await mongoDirect.stop()
    server.log('info', 'Config server stopped')
  }

  async function start() {
    await server.start()
    await database.start()
    await mongoDirect.start()
    server.log('info', `Config server started on ${env.HOST}:${env.PORT}`)
  }

  return { server, start, stop }
}
