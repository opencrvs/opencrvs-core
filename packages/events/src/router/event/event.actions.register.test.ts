/* eslint-disable max-lines */
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import { HttpResponse, http } from 'msw'
import {
  ActionStatus,
  ActionType,
  AddressType,
  createPrng,
  generateRegistrationNumber,
  getOrThrow,
  getUUID,
  SCOPES
} from '@opencrvs/commons'
import {
  createEvent,
  createTestClient,
  setupTestCase
} from '@events/tests/utils'
import { mswServer } from '@events/tests/msw'
import { env } from '@events/environment'

test('prevents forbidden access if missing required scope', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [])

  await expect(
    client.event.actions.register.request(
      generator.event.actions.register('event-test-id-12345')
    )
  ).rejects.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test(`allows access if required scope is present`, async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [SCOPES.RECORD_REGISTER])

  await expect(
    client.event.actions.register.request(
      generator.event.actions.register('event-test-id-12345')
    )
  ).rejects.not.toMatchObject(new TRPCError({ code: 'FORBIDDEN' }))
})

test('can not register an event that is not first declared and validated', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)
  const event = await createEvent(client, generator, [])

  await expect(
    client.event.actions.register.request(
      generator.event.actions.register(event.id)
    )
  ).rejects.matchSnapshot()
})

test('Validation error message contains all the offending fields', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)
  const event = await client.event.create(generator.event.create())

  await client.event.actions.declare.request(
    generator.event.actions.declare(event.id)
  )

  const createAction = event.actions.filter(
    (action) => action.type === ActionType.CREATE
  )

  const assignmentInput = generator.event.actions.assign(event.id, {
    assignedTo: createAction[0].createdBy
  })

  await client.event.actions.assignment.assign(assignmentInput)

  await client.event.actions.validate.request(
    generator.event.actions.validate(event.id)
  )

  /** Partial payload is accepted, so it should not complain about fields already send during declaration. */
  const data = generator.event.actions.register(event.id, {
    declaration: {
      'applicant.dob': '02-02',
      'recommender.none': true,
      'applicant.dobUnknown': false
    }
  })

  await client.event.actions.assignment.assign({
    ...assignmentInput,
    transactionId: getUUID()
  })
  await expect(
    client.event.actions.register.request(data)
  ).rejects.matchSnapshot()
})

test('when mandatory field is invalid, conditional hidden fields are still skipped', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)
  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.register(event.id, {
    declaration: {
      'applicant.dob': '02-1-2024',
      'applicant.dobUnknown': false,
      'applicant.name': {
        firstname: 'John',
        surname: 'Doe'
      },
      'recommender.none': true,
      'applicant.address': {
        country: 'FAR',
        addressType: AddressType.DOMESTIC,
        province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
        district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
        urbanOrRural: 'RURAL' as const,
        village: 'Small village'
      }
    }
  })

  await expect(
    client.event.actions.register.request(data)
  ).rejects.matchSnapshot()
})

const declaration = {
  'applicant.name': {
    firstname: 'John',
    surname: 'Doe'
  },
  'recommender.none': true,
  'applicant.address': {
    country: 'FAR',
    addressType: AddressType.DOMESTIC,
    province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
    district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
    urbanOrRural: 'RURAL' as const,
    village: 'Small village'
  }
}

test('Skips required field validation when they are conditionally hidden', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)
  const { id: eventId } = await createEvent(client, generator, [
    ActionType.DECLARE,
    ActionType.VALIDATE
  ])

  const data = generator.event.actions.register(eventId)

  const response = await client.event.actions.register.request(data)

  const savedAction = response.actions.find(
    (action) => action.type === ActionType.REGISTER
  )

  expect(savedAction).toMatchObject({
    status: ActionStatus.Accepted,
    declaration: data.declaration
  })
})

test('Prevents adding birth date in future', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)
  const event = await client.event.create(generator.event.create())

  const form = {
    'applicant.dob': '2040-02-01',
    'applicant.dobUnknown': false,
    'applicant.name': {
      firstname: 'John',
      surname: 'Doe'
    },
    'recommender.none': true,
    'applicant.address': {
      country: 'FAR',
      addressType: AddressType.DOMESTIC,
      province: 'a45b982a-5c7b-4bd9-8fd8-a42d0994054c',
      district: '5ef450bc-712d-48ad-93f3-8da0fa453baa',
      urbanOrRural: 'RURAL' as const,
      village: 'Small village'
    }
  }

  const payload = generator.event.actions.register(event.id, {
    declaration: form
  })

  await expect(
    client.event.actions.register.request(payload)
  ).rejects.matchSnapshot()
})

const MOCK_REGISTRATION_NUMBER = '1MY2TEST3NRO'

describe('Request and confirmation flow', () => {
  const prng = createPrng(1046)
  let registrationNumber: string

  function mockNotifyApi(status: number) {
    return mswServer.use(
      http.post<never, { actionId: string }>(
        `${env.COUNTRY_CONFIG_URL}/events/tennis-club-membership/actions/REGISTER`,
        () => {
          registrationNumber = generateRegistrationNumber(prng)
          const responseBody = status === 200 ? { registrationNumber } : {}
          // @ts-expect-error - "For some reason the msw types here complain about the status, even though this is correct"
          return HttpResponse.json(responseBody, { status })
        }
      )
    )
  }

  test('should be able to successfully call action request multiple times, without creating duplicate request actions', async () => {
    const { user, generator } = await setupTestCase()
    const client = createTestClient(user)
    const originalEvent = await createEvent(client, generator, [
      ActionType.DECLARE,
      ActionType.VALIDATE
    ])

    const { id: eventId } = originalEvent
    mockNotifyApi(200)

    const createAction = originalEvent.actions.filter(
      (action) => action.type === ActionType.CREATE
    )

    const assignmentInput = generator.event.actions.assign(originalEvent.id, {
      assignedTo: createAction[0].createdBy
    })

    await client.event.actions.assignment.assign(assignmentInput)

    const data = generator.event.actions.register(eventId, {
      declaration,
      keepAssignment: true
    })

    await client.event.actions.register.request(data)
    await client.event.actions.register.request(data)
    const event = await client.event.actions.register.request(data)

    const registerActions = event.actions.filter(
      (action) => action.type === ActionType.REGISTER
    )

    expect(registerActions).toHaveLength(1)
  })

  describe('Synchronous confirmation flow', () => {
    test('should mark action as accepted if notify API returns HTTP 200', async () => {
      const { user, generator } = await setupTestCase()
      const client = createTestClient(user)
      const { id: eventId } = await createEvent(client, generator, [
        ActionType.DECLARE,
        ActionType.VALIDATE
      ])

      mockNotifyApi(200)

      const data = generator.event.actions.register(eventId)

      const response = await client.event.actions.register.request(data)
      const savedAction = response.actions.find(
        (action) => action.type === ActionType.REGISTER
      )

      expect(savedAction).toMatchObject({
        status: ActionStatus.Accepted,
        declaration: data.declaration,
        registrationNumber: registrationNumber
      })
    })

    test('should not save action if notify API returns invalid registration number', async () => {
      const { user, generator } = await setupTestCase()
      const client = createTestClient(user)
      const { id: eventId } = await createEvent(client, generator, [
        ActionType.DECLARE,
        ActionType.VALIDATE
      ])

      mswServer.use(
        http.post(
          `${env.COUNTRY_CONFIG_URL}/events/tennis-club-membership/actions/REGISTER`,
          () => {
            return HttpResponse.json(
              { registrationNumber: 1234567890 }, // Registration number is not a string as it should be
              { status: 200 }
            )
          }
        )
      )

      const data = generator.event.actions.register(eventId)

      await expect(
        client.event.actions.register.request(data)
      ).rejects.matchSnapshot()

      const event = await client.event.get(eventId)
      const registerActions = event.actions.filter(
        (action) => action.type === ActionType.REGISTER
      )
      expect(registerActions).toHaveLength(0)
    })

    test('should mark action as rejected if notify API returns HTTP 400', async () => {
      const { user, generator } = await setupTestCase()
      const client = createTestClient(user)
      const { id: eventId } = await createEvent(client, generator, [
        ActionType.DECLARE,
        ActionType.VALIDATE
      ])

      mockNotifyApi(400)

      const data = generator.event.actions.register(eventId, {
        declaration
      })

      const response = await client.event.actions.register.request(data)
      const savedAction = response.actions.find(
        (action) => action.type === ActionType.REGISTER
      )

      expect(savedAction).toMatchObject({
        status: ActionStatus.Rejected,
        declaration
      })
    })

    test('should not save action if notify API returns HTTP 500', async () => {
      const { user, generator } = await setupTestCase()
      const client = createTestClient(user)

      const { id: eventId } = await createEvent(client, generator, [
        ActionType.DECLARE,
        ActionType.VALIDATE
      ])

      mockNotifyApi(500)

      await expect(
        client.event.actions.register.request(
          generator.event.actions.register(eventId)
        )
      ).rejects.matchSnapshot()

      const registeredEvent = await client.event.get(eventId)

      const registerActions = registeredEvent.actions.filter(
        (action) => action.type === ActionType.REGISTER
      )
      expect(registerActions).toHaveLength(0)
    })
  })

  describe('Asynchronous confirmation flow', () => {
    test('should save action in requested state if notify API returns HTTP 202', async () => {
      const { user, generator } = await setupTestCase()
      const client = createTestClient(user)

      const event = await createEvent(client, generator, [
        ActionType.DECLARE,
        ActionType.VALIDATE
      ])

      mockNotifyApi(202)

      const registerInput = generator.event.actions.register(event.id)

      const response =
        await client.event.actions.register.request(registerInput)

      const savedAction = response.actions.find(
        (action) => action.type === ActionType.REGISTER
      )

      expect(savedAction).toMatchObject({
        status: ActionStatus.Requested,
        declaration: registerInput.declaration
      })
    })

    describe('Accepting', () => {
      test('should not be able to accept the action if action is not first requested', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)
        const event = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        mockNotifyApi(202)

        const data = generator.event.actions.register(event.id, {
          declaration
        })

        await expect(
          client.event.actions.register.accept({
            ...data,
            actionId: getUUID(),
            registrationNumber: MOCK_REGISTRATION_NUMBER
          })
        ).rejects.matchSnapshot()
      })

      test('should not be able to accept action if action is already rejected', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)

        const originalEvent = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        const { id: eventId } = originalEvent
        mockNotifyApi(202)

        const data = generator.event.actions.register(eventId)

        const registerResponse =
          await client.event.actions.register.request(data)

        const originalActionId = getOrThrow(
          registerResponse.actions.find(
            (action) => action.type === ActionType.REGISTER
          )?.id,
          'Could not find register action for id'
        )

        const createAction = originalEvent.actions.filter(
          (action) => action.type === ActionType.CREATE
        )

        const assignmentInput = generator.event.actions.assign(
          originalEvent.id,
          {
            assignedTo: createAction[0].createdBy
          }
        )

        await client.event.actions.assignment.assign(assignmentInput)

        await client.event.actions.register.reject({
          eventId,
          actionId: originalActionId,
          transactionId: getUUID()
        })

        await client.event.actions.assignment.assign({
          ...assignmentInput,
          transactionId: getUUID()
        })

        await expect(
          client.event.actions.register.accept({
            ...data,
            actionId: originalActionId,
            registrationNumber: MOCK_REGISTRATION_NUMBER
          })
        ).rejects.matchSnapshot()
      })

      test('should successfully accept a previously requested action', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)

        const originalEvent = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        const { id: eventId } = originalEvent
        mockNotifyApi(202)

        const data = generator.event.actions.register(eventId)

        const registerResponse =
          await client.event.actions.register.request(data)

        const originalActionId = getOrThrow(
          registerResponse.actions.find(
            (action) => action.type === ActionType.REGISTER
          )?.id,
          'Could not find register action for id'
        )

        const createAction = originalEvent.actions.filter(
          (action) => action.type === ActionType.CREATE
        )

        const assignmentInput = generator.event.actions.assign(
          originalEvent.id,
          {
            assignedTo: createAction[0].createdBy
          }
        )
        await client.event.actions.assignment.assign(assignmentInput)

        await client.event.actions.register.request(data)

        const response = await client.event.actions.register.accept({
          ...data,
          transactionId: getUUID(),
          actionId: originalActionId,
          registrationNumber: MOCK_REGISTRATION_NUMBER
        })

        const registerActions = response.actions.filter(
          (action) =>
            action.type === ActionType.REGISTER &&
            action.status !== ActionStatus.Rejected
        )

        expect(registerActions.length).toBe(2)
        expect(registerActions[0].status).toEqual(ActionStatus.Requested)
        expect(registerActions[1]).toMatchObject({
          status: ActionStatus.Accepted,
          declaration: data.declaration,
          registrationNumber: MOCK_REGISTRATION_NUMBER,
          originalActionId: originalActionId
        })
      })

      test('should be able to call accept multiple times, without creating duplicate accept actions', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)
        const originalEvent = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        const { id: eventId } = originalEvent

        mockNotifyApi(202)

        const data = generator.event.actions.register(eventId)

        const registerResponse =
          await client.event.actions.register.request(data)

        const originalActionId = getOrThrow(
          registerResponse.actions.find(
            (action) => action.type === ActionType.REGISTER
          )?.id,
          'Could not find register action for id'
        )

        const createAction = originalEvent.actions.filter(
          (action) => action.type === ActionType.CREATE
        )

        const assignmentInput = generator.event.actions.assign(
          originalEvent.id,
          {
            assignedTo: createAction[0].createdBy
          }
        )

        await client.event.actions.assignment.assign(assignmentInput)

        await client.event.actions.register.accept({
          ...data,
          transactionId: getUUID(),
          actionId: originalActionId,
          registrationNumber: MOCK_REGISTRATION_NUMBER
        })

        await client.event.actions.assignment.assign({
          ...assignmentInput,
          transactionId: getUUID()
        })
        const response = await client.event.actions.register.accept({
          ...data,
          transactionId: getUUID(),
          actionId: originalActionId,
          registrationNumber: MOCK_REGISTRATION_NUMBER
        })

        const registerActions = response.actions.filter(
          (action) =>
            action.type === ActionType.REGISTER &&
            action.status !== ActionStatus.Rejected
        )

        expect(registerActions.length).toBe(2)
        expect(registerActions[0].status).toEqual(ActionStatus.Requested)
        expect(registerActions[1]).toMatchObject({
          declaration: data.declaration,
          status: ActionStatus.Accepted
        })
      })
      test.todo('should be able to edit the event data while accept action')
    })

    describe('Rejecting', () => {
      test('should not be able to reject the action if action is not first requested', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)
        const event = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        mockNotifyApi(202)

        const data = generator.event.actions.register(event.id, {
          declaration
        })

        await expect(
          client.event.actions.register.reject({
            ...data,
            actionId: getUUID()
          })
        ).rejects.matchSnapshot()
      })

      test('should not be able to reject the action if action is already accepted', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)
        const event = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])
        const eventId = event.id

        mockNotifyApi(202)

        const data = generator.event.actions.register(eventId)

        const registerResponse =
          await client.event.actions.register.request(data)

        const createAction = event.actions.filter(
          (action) => action.type === ActionType.CREATE
        )

        const assignmentInput = generator.event.actions.assign(event.id, {
          assignedTo: createAction[0].createdBy
        })

        const originalActionId = getOrThrow(
          registerResponse.actions.find(
            (action) => action.type === ActionType.REGISTER
          )?.id,
          'Could not find register action for id'
        )

        await client.event.actions.assignment.assign(assignmentInput)

        await client.event.actions.register.accept({
          ...data,
          actionId: originalActionId,
          transactionId: getUUID(),
          registrationNumber: MOCK_REGISTRATION_NUMBER
        })

        await client.event.actions.assignment.assign({
          ...assignmentInput,
          transactionId: getUUID()
        })
        await expect(
          client.event.actions.register.reject({
            ...data,
            actionId: originalActionId
          })
        ).rejects.matchSnapshot()
      })

      test('should be able to call reject multiple times, without creating duplicate reject actions', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)
        const event = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        const { id: eventId } = event

        mockNotifyApi(202)

        const data = generator.event.actions.register(eventId, {
          declaration
        })

        const registerResponse =
          await client.event.actions.register.request(data)

        const originalActionId = getOrThrow(
          registerResponse.actions.find(
            (action) => action.type === ActionType.REGISTER
          )?.id,
          'Could not find register action for id'
        )

        const createAction = event.actions.filter(
          (action) => action.type === ActionType.CREATE
        )

        const assignmentInput = generator.event.actions.assign(event.id, {
          assignedTo: createAction[0].createdBy
        })

        await client.event.actions.assignment.assign(assignmentInput)

        await client.event.actions.register.reject({
          eventId,
          transactionId: getUUID(),
          actionId: originalActionId
        })

        await client.event.actions.assignment.assign({
          ...assignmentInput,
          transactionId: getUUID()
        })
        const response = await client.event.actions.register.reject({
          eventId,
          transactionId: getUUID(),
          actionId: originalActionId
        })

        const registerActions = response.actions.filter(
          (action) => action.type === ActionType.REGISTER
        )

        expect(registerActions.length).toBe(2)
        expect(registerActions[0].status).toEqual(ActionStatus.Requested)
        expect(registerActions[1].status).toEqual(ActionStatus.Rejected)
      })

      test('should successfully reject a previously requested action', async () => {
        const { user, generator } = await setupTestCase()
        const client = createTestClient(user)
        const event = await createEvent(client, generator, [
          ActionType.DECLARE,
          ActionType.VALIDATE
        ])

        const { id: eventId } = event
        mockNotifyApi(202)

        const data = generator.event.actions.register(eventId)

        const registerResponse =
          await client.event.actions.register.request(data)

        const originalActionId = getOrThrow(
          registerResponse.actions.find(
            (action) => action.type === ActionType.REGISTER
          )?.id,
          'Could not find register action for id'
        )

        const createAction = event.actions.filter(
          (action) => action.type === ActionType.CREATE
        )

        const assignmentInput = generator.event.actions.assign(event.id, {
          assignedTo: createAction[0].createdBy
        })

        await client.event.actions.assignment.assign(assignmentInput)

        const response = await client.event.actions.register.reject({
          eventId,
          transactionId: getUUID(),
          actionId: originalActionId
        })

        const registerActions = response.actions.filter(
          (action) => action.type === ActionType.REGISTER
        )

        expect(registerActions.length).toBe(2)
        expect(registerActions[0].status).toEqual(ActionStatus.Requested)
        expect(registerActions[1]).toMatchObject({
          status: ActionStatus.Rejected,
          originalActionId
        })
      })
    })
  })
})
