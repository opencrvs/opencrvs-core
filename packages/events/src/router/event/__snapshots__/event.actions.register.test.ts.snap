// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Prevents adding birth date in future 1`] = `[TRPCError: [{"message":"Please enter a valid date","id":"applicant.dob","value":"2040-02-01"}]]`;

exports[`Request and confirmation flow > Asynchronous confirmation flow > Accepting > should not be able to accept action if action is already rejected 1`] = `[TRPCError: Action has already been rejected.]`;

exports[`Request and confirmation flow > Asynchronous confirmation flow > Accepting > should not be able to accept the action if action is not first requested 1`] = `[TRPCError: Action not found.]`;

exports[`Request and confirmation flow > Asynchronous confirmation flow > Rejecting > should not be able to reject the action if action is already accepted 1`] = `[TRPCError: Action has already been accepted.]`;

exports[`Request and confirmation flow > Asynchronous confirmation flow > Rejecting > should not be able to reject the action if action is not first requested 1`] = `[TRPCError: Action not found.]`;

exports[`Request and confirmation flow > Synchronous confirmation flow > should not save action if notify API returns HTTP 500 1`] = `[TRPCError: Unexpected failure from notification API]`;

exports[`Request and confirmation flow > Synchronous confirmation flow > should not save action if notify API returns invalid registration number 1`] = `[TRPCError: Invalid payload received from notification API]`;

exports[`Validation error message contains all the offending fields 1`] = `[TRPCError: [{"message":"Invalid date field","id":"applicant.dob","value":"02-02"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-02"}]]`;

exports[`when mandatory field is invalid, conditional hidden fields are still skipped 1`] = `[TRPCError: [{"message":"Invalid date field","id":"applicant.dob","value":"02-1-2024"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-1-2024"}]]`;
